### DISCUSSION TITLE
Python Expression API: Overview to track coverage

### DISCUSSION THREAD

**Discussion by Tishj** [duckdb/duckdb#14725] [General]
## Expressions

Generated using a variation of the script shameless stolen from <https://github.com/duckdb/duckdb/discussions/14525>
<details>
<summary>Script used to generate the table</summary>

(the "Implemented in Python" tab was manually adjusted after generation)
```python
import os
import pandas as pd
import duckdb

def to_pascal_case(snake_str):
    """Converts a snake_case string to PascalCase."""
    return ''.join(word.capitalize() for word in snake_str.split('_'))

def find_duckdb_expressions(directory):
    return [
        (to_pascal_case(filename.replace('.cpp', '')), os.path.join(directory, filename))
        for filename in os.listdir(directory)
        if filename.endswith('_expression.cpp')
    ]

DUCKDB_PARSED_EXPRESSIONS_PATH = 'src/parser/expression'

all_expressions = find_duckdb_expressions(DUCKDB_PARSED_EXPRESSIONS_PATH)

def get_implemented_types():
    implemented = [x for x in list(duckdb.__dir__()) if x.endswith('Expression') and x != 'Expression']
    mapping = {
        'Column': 'Columnref'
    }
    result = []
    for x in implemented:
        type = x[0:-10]
        result.append(f'{mapping.get(type, type)}Expression')
    return result

implemented = get_implemented_types()
raw_coverage = [(x[0], x[0] in implemented, f'[source](https://github.com/duckdb/duckdb/tree/main/{x[1]})') for x in all_expressions]

coverage = pd.DataFrame(raw_coverage, columns=["Expression", "Implemented in Python", "Path to source file"])
coverage = coverage.drop_duplicates()

print("Coverage:", format(coverage["Implemented in Python"].mean(), ".0%"))
coverage["Implemented in Python"] = coverage["Implemented in Python"].map({True: "✅", False: "❌"})

coverage.sort_values("Expression").to_markdown("tmp/coverage.md", index=False)

```

</details>

| Expression                    | Implemented in Python   | Path to source file                                       |
|:------------------------------|:------------------------|:----------------------------------------------------------|
| BetweenExpression             | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/between_expression.cpp)              |
| CaseExpression                | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/case_expression.cpp)                 |
| CastExpression                | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/cast_expression.cpp)                 |
| CollateExpression             | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/collate_expression.cpp)              |
| ColumnrefExpression           | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/columnref_expression.cpp)            |
| ComparisonExpression          | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/comparison_expression.cpp)           |
| ConjunctionExpression         | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/conjunction_expression.cpp)          |
| ConstantExpression            | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/constant_expression.cpp)             |
| DefaultExpression             | ❌                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/default_expression.cpp)              |
| FunctionExpression            | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/function_expression.cpp)             |
| LambdaExpression              | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/lambda_expression.cpp)               |
| LambdarefExpression           | ❌                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/lambdaref_expression.cpp)            |
| OperatorExpression            | ✅*                     | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/operator_expression.cpp)             |
| ParameterExpression           | ❌                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/parameter_expression.cpp)            |
| PositionalReferenceExpression | ❌                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/positional_reference_expression.cpp) |
| StarExpression                | ✅                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/star_expression.cpp)                 |
| SubqueryExpression            | ❌                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/subquery_expression.cpp)             |
| WindowExpression              | ❌                      | [source](https://github.com/duckdb/duckdb/tree/main/src/parser/expression/window_expression.cpp)               |

There is some more nuance to this list, not everything might be *fully* implemented, one example is OperatorExpression which is probably missing some support as of writing.

If there is missing functionality but the item is marked as completed, please leave a comment addressing this so we can be aware of the limitation 👍

**Comment by Tishj** [comment]
Wrote up a python definition for a WindowExpression, it's quite a complex expression to formulate
```py
class WindowExclude:
    CURRENT_ROW: WindowExclude
    GROUP: WindowExclude
    TIES: WindowExclude
    NO_OTHERS: WindowExclude
    def __int__(self) -> int: ...
    def __index__(self) -> int: ...
    @property
    def __members__(self) -> Dict[str, WindowExclude]: ...
    @property
    def name(self) -> str: ...
    @property
    def value(self) -> int: ...

class WindowSpecType:
    RANGE: WindowSpecType
    ROWS: WindowSpecType
    def __int__(self) -> int: ...
    def __index__(self) -> int: ...
    @property
    def __members__(self) -> Dict[str, WindowSpecType]: ...
    @property
    def name(self) -> str: ...
    @property
    def value(self) -> int: ...


class WindowBound:
    CURRENT_ROW: WindowBound
    UNBOUNDED: WindowBound
    PRECEDING: WindowBound
    FOLLOWING: WindowBound
    def __int__(self) -> int: ...
    def __index__(self) -> int: ...
    @property
    def __members__(self) -> Dict[str, WindowBound]: ...
    @property
    def name(self) -> str: ...
    @property
    def value(self) -> int: ...

class WindowSpec:
	def __init__(
			self,
			type: WindowSpecType,
			start: Union[WindowBound, WindowBoundWithExpression],
			end: Optional[Union[WindowBound, WindowBoundWithExpression]] = None
		) -> WindowSpec: ...

def WindowExpression(
		function_name: str,
		*args: "Expression",
		distinct: bool = False,
		ignore_nulls: bool = False,
		filter: "Expression" = None,
		partition_by: List["Expression"] = [],
		order_by: List["Expression"] = [],
		window_spec: Optional[WindowSpec] = None,
		exclude: Optional[WindowExclude] = None
)
```

Not entirely sure how yet, but I'd like to be able to use `CURRENT_ROW` or `PRECEDING(expr)` (same for `FOLLOWING(expr)`)
I think that makes for a *decent* API, hopefully

Which means `WindowBound` needs to have a `__call__` method, only accepted on PRECEDING and FOLLOWING, that take an Expression and return a WindowBoundWithExpression

**Reply by Tishj** [reply]
Might want to take a `function: FunctionExpression` instead of `function_name: str`
One reason for that is that the function itself could have an `ORDER BY`

Though the `ORDER BY` for the window spec doesn't seem to be implemented currently:
```
Parser Error: ORDER BY is not implemented for window functions!
```

**Reply by binste** [reply]
Exciting stuff! :) Feel free to ping me once this lands and I can take over the implementation of the various window functions in the PySpark API.

**Reply by MarcoGorelli** [reply]
Is the idea that then users would be able to write calls such as:
```python
rel: duckdb.DuckDBPyRelation

expr_1 = duckdb.WindowExpression(
    duckdb.FunctionExpression('sum', duckdb.ColumnExpression('a')),
)
rel.select(expr_1)
expr_2 = duckdb.WindowExpression(
    duckdb.FunctionExpression('sum', duckdb.ColumnExpression('a')),
    order_by=duckdb.ColumnExpression('b'),
)
rel.select(expr_2)
```
?

If so, this sounds wonderful 😍

**Reply by binste** [reply]
I've just updated the overview in #14525 and I think this is one of the main remaining missing pieces. @Tishj Friendly ping to check in if you have visibility on your side if this is something you will have time to work on or if there are other priorities right now? Happy to take over the implementation of the various window functions in the PySpark API once this lands.

PS: Great to see `expr` based on `SQLExpression`! Super useful :)

**Comment by Tishj** [comment]
Also related: <https://github.com/duckdb/duckdb/discussions/14936>

**Comment by Tishj** [comment]
https://github.com/duckdb/duckdb/discussions/15485

**Comment by Tishj** [comment]
https://github.com/duckdb/duckdb/discussions/15721

sounds like this should also be tracked here

**Comment by MarcoGorelli** [comment]
A few more which haven't been mentioned here yet:

- https://github.com/duckdb/duckdb/discussions/16996
- https://github.com/duckdb/duckdb/discussions/16947
- https://github.com/duckdb/duckdb/discussions/16763

**Comment by djouallah** [comment]
Read Delta https://github.com/duckdb/duckdb-delta/issues/175

**Reply by Tishj** [reply]
Not the place
